histogram/src/histogram.rs

```diff
@@ -0,0 +1,430 @@
+        // concurrent modification that reduces the counts before we have a
+        // chance to get to that bucket
+        Ok(self.get_bucket(max))
+    }
+
+    /// Returns a set of percentiles in a single and efficient bulk operation.
+    /// Note that the returned percentiles will be sorted from lowest to highest
+    /// in the result, even if they do not appear in that order in the provided
+    /// set of requested percentiles.
+    pub fn percentiles(&self, percentiles: &[f64]) -> Result<Vec<Percentile>, Error> {
+        let mut percentiles = percentiles.to_owned();
+        percentiles.sort_by(|a, b| a.partial_cmp(b).unwrap());
+
+        for percentile in &percentiles {
+            if !(0.0..=100.0).contains(percentile) {
+                return Err(Error::InvalidPercentile);
+            }
+        }
+
+        let total: u64 = self
+            .buckets
+            .iter()
+            .map(|v| v.load(Ordering::Relaxed) as u64)
+            .sum();
+        if total == 0 {
+            return Err(Error::Empty);
+        }
+
+        let thresholds: Vec<u64> = percentiles
+            .iter()
+            .map(|v| std::cmp::min(1, (v * total as f64 / 100.0).ceil() as u64))
+            .collect();
+
+        let mut max = 0;
+        let mut seen = 0;
+        let mut threshold_idx = 0;
+
+        let mut result = Vec::with_capacity(thresholds.len());
+
+        for (bucket_idx, count) in self
+            .buckets
+            .iter()
+            .map(|b| b.load(Ordering::Relaxed) as u64)
+            .enumerate()
+        {
+            if count > 0 {
+                max = bucket_idx;
+            }
+
+            seen += count;
+            while seen > thresholds[threshold_idx] && threshold_idx < thresholds.len() {
+                result.push(Percentile {
+                    percentile: percentiles[threshold_idx],
+                    bucket: self.get_bucket(bucket_idx),
+                });
+
+                threshold_idx += 1;
+            }
+
+            if threshold_idx >= thresholds.len() {
+                break;
+            }
+        }
+
+        // pad the results with the max bucket seen while walking the histogram
+        // this may be necessary if there is a concurrent modification that
+        // reduces the counts before we have a chance to get to that bucket
+        while result.len() < percentiles.len() {
+            // get the index within the percentiles vec
+            let idx = percentiles.len() - result.len() - 1;
+            result.push(Percentile {
+                percentile: percentiles[idx],
+                bucket: self.get_bucket(max),
+            });
+        }
+
+        Ok(result)
+    }
+
+    /// Adds the other `Histogram` to this `Histogram`. Returns an error if
+    /// there are differences in the configurations of both `Histogram`s.
+    #[allow(clippy::result_unit_err)]
+    pub fn add_assign(&self, other: &Self) -> Result<(), ()> {
```

Comment:
It would be a good idea to create an error type for this. `Result<(), ()>` doesn't play well with libraries like anyhow. Something like this would work (if I remember my thiserror right)

```rs
#[derive(Error)]
#[error("histograms had incompatible parameters")]
struct IncompatibleHistogramError;
```

Also my impression is that most of the literature calls this operation `merge`, up to you on which name you like better.

Reply:
Is there a good equivalent for subtraction to go with merge? I really wanted to use the SubAssign and AddAssign traits, but they're not a good fit for this. 

We already have an error type, so adding something more useful here should be trivial.

Reply:
What about `subtract`? It's a verb so IMO it's better than `sub_assign` and the underlying operation is set difference.

Reply:
👍 works for me.